File: Div3-Group1/ReverseLinkedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head,ListNode* AR,int l, int r) {
        if(head==NULL || head->next==NULL)
            return head;
            
        ListNode* present=head;
        ListNode* Next=head->next;
        ListNode* preve=AR;

        while(present->next!=NULL && l++ < r)
        {
            present->next=preve;
            preve=present;

            if(Next!=NULL)
                present=Next;
            if(Next->next!=NULL)
                Next=Next->next;
            else
                Next==NULL;
        }

            present->next=preve;
            preve=present;

        return preve;
    }

    ListNode* reverseBetween(ListNode* head, int left, int right) {
        if(head==NULL || head->next==NULL || right-left==0)
            return head;

        ListNode* temp=head;
        ListNode* present=head;
        ListNode* Next=head->next;

        ListNode* Interval=NULL;

        ListNode* AR=NULL;

        int l=1;

        while(l++ < left-1)
            temp=temp->next;
        
        AR=temp->next;
        
        while(l++ < right)
            AR=AR->next;
        
        if(AR -> next)
            AR=AR->next;
        else
            AR=NULL;


        if(left==1)
        {
            Interval = reverseList(temp,AR,left,right);
            return Interval;
        }
        else
        {
            Interval = reverseList(temp->next,AR,left,right);
        }

        
        temp->next = Interval;

        // cout<<AR->val;

        // Interval->next=AR;
    
       
       return head;
    }
};

File: leet-code-solutions/leetcode/reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        slow = head
        new_head = ListNode(None, None)
        new_head.next = head
        left_edge = ListNode(None, None)
        left_head = ListNode(None, None)
        prev = ListNode(None, None)
        trans = ListNode(None, None)
        
        if head.next is None or left == right:
            return head

        if left == 1:
            slow = new_head
            left_edge = slow
            for i in range(left-1):
                left_edge = slow
                slow=slow.next
            slow=slow.next
            left_head = slow
            
            trans=slow.next
            while (right-left):
                prev = slow
                slow = trans
                trans = slow.next
                slow.next = prev
                right-=1
            
            left_edge.next=slow
            left_head.next=trans

            return new_head.next

        left_edge = slow
        for i in range(left-2):
            slow=slow.next
            left_edge = slow
        slow=slow.next
        left_head = slow
        
        trans=slow.next
        while (right-left):
            prev = slow
            slow = trans
            trans = slow.next
            slow.next = prev
            right-=1
        
        left_edge.next=slow
        left_head.next=trans

        return head
